src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -0,0 +1,76 @@
+from collections import defaultdict
+
+import torch as th
+
+from poprox_concepts import Article, ArticleSet, InterestProfile
+from poprox_recommender.diversifiers import Calibrator, compute_kl_divergence
+from poprox_recommender.lkpipeline import Component
+from poprox_recommender.topics import extract_locality, normalized_topic_count
+
+
+# Locality Calibration uses MMR
+# to rerank recommendations according to
+# locality calibration
+class LocalityCalibrator(Component, Calibrator):
+    def __init__(self, theta: float = 0.1, num_slots=10):
+        super.__init__(theta, num_slots)
+
+    def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestProfile) -> ArticleSet:
+        normalized_locality_prefs = normalized_topic_count(interest_profile.click_locality_counts)
+
+        if candidate_articles.scores is not None:
+            article_scores = th.sigmoid(th.tensor(candidate_articles.scores))
+        else:
+            article_scores = th.zeros(len(candidate_articles.articles))
+
+        article_scores = article_scores.cpu().detach().numpy()
+
+        article_indices = self.calibration(
+            article_scores,
+            candidate_articles.articles,
+            normalized_locality_prefs,
+            self.theta,
+            topk=self.num_slots,
+        )
+        return ArticleSet(articles=[candidate_articles.articles[int(idx)] for idx in article_indices])
+
+    def calibration(self, relevance_scores, articles, preferences, theta, topk) -> list[Article]:
+        # MR_i = \theta * reward_i - (1 - \theta)*C(S + i) # C is calibration
+        # R is all candidates (not selected yet)
+
+        recommendations = []  # final recommendation (topk index)
+        rec_categories = defaultdict(int)  # frequency distribution of catregories of S
```

Comment:
Not sure if I understand this comment. Maybe its just copied from topic_calibration?

Reply:
I think the `S` vs `R` nomenclature comes from the MMR paper, which then got carried into our calibrator implementation. `R` is the set of available candidates Remaining; `S` is the set of items Selected so far.